jbwm.1
.TH jbwm 1 "01 JAN 2017"

.SH NAME
jbwm \- minimalistic window manager for X

.SH SYNOPSIS
.B jbwm
[\fB-1\fR \fIMODS\fR]
[\fB-2\fR \fIMODS\fR]
[\fB-v\fR]

.SH DESCRIPTION
.B jbwm
is a minimalist window manager forked from evilwm, adding basic title bar
support, simpler key bindings, and various code cleanups and size
reductions.  evilwm is based on aewm, extended to feature
many keyboard controls, and otherwise altered to be more friendly.

.SH OPTIONS
.IP "\fB-1 \fIMOD"
Set first modifier key.  \fBMOD\fR may be control, alt, mod1, mod2, mod3,
mod4, or mod5.
.IP "\fB-2 \fIMOD"
Set second modifier key.  \fBMOD\fR is as above.
.IP "\fB-v"
Print version then exit.
.IP "\fB-*"
For all other options, display usage then exit.

.SH NOTES
In jbwm, the focus follows the mouse pointer, and focus is not lost if
you stray onto the root window. 
.PP
You can use the mouse to manipulate windows either by click/dragging
the 1 pixel border, or by holding down \fBMod1\fR and doing so anywhere 
in the client window. The controls are:
.IP "\fBButton 1"
Move or raise window.
.IP "\fBButton 2"
Lower window.
.IP "\fBButton 3"
Resize window.
.PP
Most keyboard controls are used by holding down \fBMod1\fR, then
pressing a key. Available functions are:
.IP \fBEscape
Exit jbwm.
.IP \fBReturn
Spawn new terminal.
.IP \fBq
Delete current window.
.IP "\fBDown Arrow"
Lower current window.
.IP "\fBUp Arrow"
Raise current window.
.IP "\fBH, J, K, L"
Move window left, down, up or right.  Holding \fBShift\fR resizes the
window instead.
.IP \fBa
Toggle fullscreen.  
.IP \fBSpace
Toggle maximization.
.IP \fBz
Toggle vertical maximization.  
.IP \fBx
Toggle horizontal maximization.  
.IP \fBs
Toggle window shade.
.PP
If compiled with virtual desktop support, these functions are also available: 
.IP \fB.
Toggle window stickiness.
.IP "\fB1--0"
Switch virtual desktop.
.IP \fBLeft
Previous virtual desktop.
.IP \fBRight
Next virtual desktop.
.PP
In addition to the above, \fBMod1+Tab\fR can be used 
to cycle through windows on screen.

.SH ENVIRONMENT
.IP \fBDISPLAY
DISPLAY should be set, otherwise default of :0 will be used.  
.IP \fBJBWM_FC
Set sticky window border color.
.IP \fBJBWM_FG
Set foreground color.
.IP \fBJBWM_BG
Set background color.
.IP \fBJBWM_CLOSE
Set close button color.
.IP \fBJBWM_SHADE
Set shade button color.
.IP \fBJBWM_RESIZE
Set resize button color.
.IP \fBJBWM_FONT
Set font.
.IP \fBJBWM_TERM
Set terminal emulator program.

.SH BUGS
EWMH support is a work in progress.  

.SH AUTHOR
Jeffrey Bedard <\fIalisabedard@gmail.com\fR>
Ciaran Anscomb <\fIevilwm@6809.org.uk\fR> wrote evilwm.
Decklin Foster <\fIfosterd@hartwick.edu\fR> wrote aewm.
David Hogan <\fIdhog@cs.su.oz.au\fR> wrote 9wm.

.SH "SEE ALSO"
.BR xinit (1),
.BR X (7)


ChangeLog

*** Changes between 1.2 and 1.10
	* Add XBM titlebar support.
	* Add XPM titlebar support.
	* Add Xft font support.
	* Fix excessive titlebar updates by correctly selecting
	  c->info_window events.  



*** Changes between 1.1 and 1.2
Thu Aug 23 15:48:21 EDT 2007 Jeffrey Bedard <jefbed@gmail.com>
  * Removed INFOBANNER support for non-TITLEBAR configurations.
    Refactored titlebar code, removed non-TITLEBAR preprocessor
    code paths.  Added virtual desktop window assignment with 
    scroll wheel.  

*** Changes between 1.0 and 1.1

Thu Aug 16 09:54:22 EDT 2007  Jeffrey Bedard <jefbed@gmail.com>
  * Fixed shrinking.  Fixed killing children.  Fixed some titlebar
    metrics.  

*** Changes between 1.0 and evilwm-1.0.0/evilwm-0.99.25

Sat Aug 11 20:07:34 EST 2007  Jeffrey Bedard <jefbed@gmail.com>
  * Decided to fork, considering code base has changed significantly,
    thus preventing seamless integration of evilwm changes.  This
    fork/branch will be called jbwm, the jefbed window manager.
    It is intended for general use, but some features may be added
    with special consideration of the jefbed scripting environment.  
    This fork so far adds titlebar support (optional), fixes keyboard
    resizing, many many code cleanups, optimizations based on profiling
    data, reduced binary size.

*** Changes between 0.99.25 and 1.0.0:

Thu May 17 15:48:31 BST 2007  Ciaran Anscomb <evilwm@6809.org.uk>
  * When snapping, prefer screen edge to other windows

Wed May  3 16:53:01 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Don't account for width_inc and height_inc when maximising

Wed May  3 15:38:08 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Ignore borders when snapping maximised client to screen edge

*** Changes between 0.99.24 and 0.99.25:

Wed Apr 26 18:35:53 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Regrab keys on receipt of MappingNotify event

Wed Apr 26 13:01:19 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Make spawn() use find_current_screen()

Wed Apr 26 11:30:04 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Fix build errors/warnings when VWM not defined

Tue Apr 25 19:35:44 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Track vdesk separately on each managed screen

Tue Apr 25 18:42:29 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Remove is_sticky tests when only need to check vdesk

Tue Apr 25 17:00:09 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Only snap against clients on the same screen

Thu Apr 13 16:15:09 BST 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Fix configure requests containing sibling information

Mon Mar 13 12:53:55 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Move unused variable workarounds after declarations

Tue Mar  7 08:13:59 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Flesh out some switches with default cases

Mon Feb 27 16:49:18 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Remove redundant call to XShapeSelectInput()

Mon Feb 27 14:52:48 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Move grab_keysym() to main.c, make static

Mon Feb 27 14:36:19 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Make recalculate_sweep() static

Mon Feb 27 14:20:58 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Wrap client_update_current into select_client

Wed Feb 22 16:40:39 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Restructure client_update_current()

Wed Feb 22 12:39:48 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Wrap fetching of mwm_hints property into init_size

Wed Feb 22 10:58:45 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Omit some unneeded code when VWM not defined
  EWMH atoms mostly, also some keymap defines.

Wed Feb 22 08:57:14 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Code tidy
  Dead declarations removed from evilwm.h.  Some reordering of other
  declarations.  Nothing that should affect anything.

Tue Feb 21 20:10:32 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Use XQueryPointer to find pointer root instead of trying to track it

Tue Feb 21 20:09:57 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Macro-ise get_mouse_position()

Tue Feb 21 19:36:08 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Declare some of opt_* as static in main.c

Tue Feb 21 16:36:26 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Don't track bounding_shaped
  This is never used outside set_shape and I'm not convinced that
  XShapeQueryExtents refers to the old values.

Tue Feb 21 16:16:46 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Fix foolish lookup of screen number where it's already known

Tue Feb 21 15:00:15 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Expose events not needed on frame windows

Tue Feb 21 14:59:31 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Grab buttons when making client instead of every enter event

Tue Feb 21 12:12:18 GMT 2006  Ciaran Anscomb <evilwm@6809.org.uk>
  * Refactored handle_key_event()



*** Old ChangeLog:
	Released version 0.3.3.

2000-10-26
	Updated aewm license info by request.
	XK_Insert does same as XK_KP_Insert - may have to rethink that as
	  it's a 4-key chord on a happy hacking.  Almost as bad as emacs!
	Alt+Tab warps account for border width

2000-10-27
	Cleared out // comments.
	Now gets the border the right size when a window resizes itself.
	First window gets an active border now.

2000-10-31
	Updated aewm license info again ;)
	Released version 0.3.4.

2000-11-06
	Some code tidyups.
	Looked into the gravity code and made it more right.

2000-11-07
	Divide XxY display by increments if bit set whilst resizing/moving.
	Released version 0.3.5.
	Shaved some int8_ts off by grabbing keys in a loop.
	-bw option to change border width.
	Also changed all other options to single - due to making more sense

2000-11-08
	Stop remembering the name of the window.
	Fixed that daft "last window's gone, can't do anything" bug
	Realised we don't care if window is transient - nothing happens anyway.

jbwm.c
// jbwm - Minimalist Window Manager for X
// Copyright 2008-2019, Alisa Bedard <alisabedard@gmail.com>
// Copyright 1999-2015, Ciaran Anscomb <evilwm@6809.org.uk>
// See README for license and other details.
//#undef DEBUG
#include "jbwm.h"
#include "config.h"
#include "ewmh.h"
#include "font.h"
#include "atom.h"
#include "JBWMScreen.h"
#include "keys.h"
#include "log.h"
#include "new.h"
#include <stdbool.h>
#include "util.h"
#include "vdesk.h"
#ifdef JBWM_USE_XFT
static void allocate_xft_color(Display * d, struct JBWMScreen * s)
{
  XftColorAllocName(d, DefaultVisualOfScreen(s->xlib),
    DefaultColormapOfScreen(s->xlib), JBWM_FG,
    &s->font_color);
}
#endif//JBWM_USE_XFT
static void allocate_colors(struct JBWMScreen * s)
{
  Display * d=s->xlib->display;
  const uint8_t n = s->id;
#define PIX(field, color) s->pixels.field = jbwm_get_pixel(d, n, color);
  PIX(bg, JBWM_BG);
  PIX(fc, JBWM_FC);
  PIX(fg, JBWM_FG);
  PIX(close, JBWM_CLOSE);
  PIX(resize, JBWM_RESIZE);
  PIX(shade, JBWM_SHADE);
  PIX(stick, JBWM_STICK);
#undef PIX
#ifdef JBWM_USE_XFT
  allocate_xft_color(d, s);
#endif//JBWM_USE_XFT
}
static bool check_redirect(Display * d, const Window w)
{
  XWindowAttributes a;
  XGetWindowAttributes(d, w, &a);
  JBWM_LOG("check_redirect(0x%x): override_redirect: %s, "
    "map_state: %s", (int)w,
    a.override_redirect ? "true" : "false",
    a.map_state == IsViewable ? "IsViewable"
    : "not IsViewable");
  return (!a.override_redirect && (a.map_state == IsViewable));
}
// Free returned data with XFree()
static Window * get_windows(Display * dpy, const Window root,
  uint16_t * win_count)
{
  Window * w, d;
  unsigned int n;
  XQueryTree(dpy, root, &d, &d, &w, &n);
  *win_count = n;
  return w;
}
static void setup_clients(Display * d, struct JBWMScreen * s,
  struct JBWMClient ** head_client, struct JBWMClient ** current_client)
{
  uint16_t n;
  Window * w = get_windows(d, RootWindow(d, s->id), &n);
  JBWM_LOG("Started with %d clients", n);
  if (w) { // Avoid segmentation fault on empty list.
    while(n--)
      if(check_redirect(d,w[n]))
        jbwm_new_client(s, head_client, current_client, w[n]);
    XFree(w);
  }
}
static inline void setup_gc(Display * d, struct JBWMScreen * s)
{
  XGCValues v={
    .foreground = s->pixels.fg,
    .background = s->pixels.bg,
    .subwindow_mode = IncludeInferiors,
    .line_width = 1
  };
  unsigned long mask = GCSubwindowMode | GCLineWidth
  | GCForeground | GCBackground;
#ifndef JBWM_USE_XFT
  v.font=s->font->fid;
  mask|=GCFont; 
#endif//JBWM_USE_XFT
  s->gc=DefaultGC(d,s->id);
  XChangeGC(d,s->gc,mask,&v);
  s->border_gc=XCreateGC(d,s->xlib->root,GCFunction,
    &(XGCValues){.function=GXxor});
  XCopyGC(d,s->gc,mask,s->border_gc);
}
static inline void setup_event_listeners(Display * d, const Window root)
{
  enum {
    EMASK = SubstructureRedirectMask | SubstructureNotifyMask |
    EnterWindowMask | PropertyChangeMask
    | ColormapChangeMask
  };
  XChangeWindowAttributes(d, root, CWEventMask,
    &(XSetWindowAttributes){.event_mask = EMASK });
}
#ifdef JBWM_USE_XFT
/* Create a unique XftDraw for each screen to properly handle colormaps and
 * screen limitations.  */
static XftDraw * new_xft_draw(Screen * s)
{
  return XftDrawCreate(DisplayOfScreen(s), RootWindowOfScreen(s),
    DefaultVisualOfScreen(s), DefaultColormapOfScreen(s));
}
#endif//JBWM_USE_XFT
// Initialize SCREENS amount of screens.
void jbwm_init_screens(Display *d, struct JBWMScreen *s,
  struct JBWMClient ** head_client, struct JBWMClient ** current_client,
  const short screens)
{
  if(screens>=0){
    unsigned long *lprop;
    uint16_t n;
    uint8_t vdesk;
    JBWM_LOG("jbwm_init_screen(d, screens), screens is %d", screens);
    s->id = screens;
    s->xlib = ScreenOfDisplay(d, screens);
    lprop=jbwm_get_property(d,s->xlib->root,
      jbwm_atoms[JBWM_NET_CURRENT_DESKTOP],&n);
    if(n){
      vdesk=lprop[0];
      XFree(lprop);
    }else
      vdesk=0;
#ifdef JBWM_USE_XFT
    s->xft = new_xft_draw(s->xlib);
#else//!JBWM_USE_XFT
    s->font = jbwm_get_font();
#endif//JBWM_USE_XFT
    s->font_height=jbwm_get_font_height();
    allocate_colors(s);
    setup_gc(d, s);
    { // r scope
      const Window r = RootWindow(d, screens);
      setup_event_listeners(d, r);
      jbwm_grab_root_keys(d, r);
    }
    /* scan all the windows on this screen */
    setup_clients(d, s, head_client, current_client);
    jbwm_ewmh_init_screen(d, s);
    jbwm_init_screens(d, s, head_client, current_client, screens - 1);
    jbwm_set_vdesk(s, *head_client, vdesk);
  }
}
